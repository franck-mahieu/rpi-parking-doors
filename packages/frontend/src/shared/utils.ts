const env = process.env.NODE_ENV;
const reactAppDevApiUrl = process.env.REACT_APP_DEV_API_URL;

export const getBaseUrl = (
  envParam: string | undefined,
  reactAppApiUrlParam: string | undefined,
) => {
  return envParam === 'development' ? reactAppApiUrlParam : '';
};

export const fetchUtils = async (uri: string, method: string, body?: any) => {
  const response = await fetch(`${getBaseUrl(env, reactAppDevApiUrl)}${uri}`, {
    method: method,
    headers: {
      'Content-Type': 'application/json',
    },
    credentials: env === 'development' ? 'include' : 'same-origin',
    body: body,
  });
  const result: any = await response.json();
  if (
    (response.status >= 200 && response.status < 300) ||
    (result?.changes && result.changes !== 0)
  ) {
    return result;
  }
  return Promise.reject(result);
};
