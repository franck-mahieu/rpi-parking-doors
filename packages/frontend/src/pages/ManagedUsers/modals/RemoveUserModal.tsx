import React, { Component } from 'react';
import M from 'materialize-css';
import 'materialize-css/dist/css/materialize.min.css';
import { fetchUtils } from '../../../shared/utils';
import './Modals.scss';

const {
  REACT_APP_LABEL_REMOVE_BUTTON,
  REACT_APP_REMOVE_USER_MODAL_TITLE,
  REACT_APP_REMOVE_USER_MODAL_DESCRIPTION,
  REACT_APP_LABEL_BUTTON_CANCEL,
  REACT_APP_LABEL_BUTTON_VALIDATE,
  REACT_APP_REMOVE_USER_SUCCESS_MESSAGE,
  REACT_APP_REMOVE_USER_ERROR_MESSAGE,
} = process.env;

interface IRemoveUserModal {
  login: string;
  guid: string;
  getUsers: any;
}

class RemoveUserModal extends Component<IRemoveUserModal> {
  private Modal: any;

  componentDidMount() {
    const options = {
      onCloseEnd: () => {
        this.props.getUsers();
      },
      inDuration: 250,
      outDuration: 250,
      opacity: 0.5,
      dismissible: true,
      startingTop: '4%',
      endingTop: '10%',
    };
    M.Modal.init(this.Modal, options);
  }

  onClickRemoveUser = async (login: string, guid: string) => {
    try {
      await fetchUtils(
        `/api/users/remove?guid=${guid}`,
        'DELETE',
        JSON.stringify({ login: login }),
      );
      M.toast({
        html: `${login} ${REACT_APP_REMOVE_USER_SUCCESS_MESSAGE}`,
        classes: 'green',
      });
    } catch (e) {
      M.toast({
        html: `${login} ${REACT_APP_REMOVE_USER_ERROR_MESSAGE}`,
        classes: 'red',
      });
    }
  };

  render() {
    return (
      <span id={'user-modal'}>
        <button
          className="waves-effect waves-light btn modal-trigger user-modal remove-button"
          data-target={`remove-user-modal-${this.props.login}`}
          title={REACT_APP_REMOVE_USER_MODAL_TITLE}
        >
          {REACT_APP_LABEL_REMOVE_BUTTON}
        </button>
        <div
          ref={Modal => {
            this.Modal = Modal;
          }}
          id={`remove-user-modal-${this.props.login}`}
          className="modal"
        >
          <div className="modal-content">
            <h4>
              {REACT_APP_REMOVE_USER_MODAL_TITLE} {this.props.login}
            </h4>
            <p>
              {REACT_APP_REMOVE_USER_MODAL_DESCRIPTION} {this.props.login} ?
            </p>
          </div>
          <div className="modal-footer">
            <button className="modal-close waves-effect waves-red btn-flat cancel-button">
              {REACT_APP_LABEL_BUTTON_CANCEL}
            </button>
            <button
              id={'remove-user-validate-button'}
              className="modal-close waves-effect waves-green btn-flat"
              onClick={async () => {
                await this.onClickRemoveUser(this.props.login, this.props.guid);
              }}
            >
              {REACT_APP_LABEL_BUTTON_VALIDATE}
            </button>
          </div>
        </div>
      </span>
    );
  }
}

export default RemoveUserModal;
