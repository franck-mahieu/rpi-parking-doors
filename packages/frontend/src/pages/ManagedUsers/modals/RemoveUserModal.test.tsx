import React from 'react';
import renderer from 'react-test-renderer';
import RemoveUserModal from './RemoveUserModal';

import * as utils from '../../../shared/utils';
import M from 'materialize-css';

jest.mock('../../../shared/utils');

describe('RemoveUserModal', () => {
  let removeUserModal: RemoveUserModal;
  beforeEach(() => {
    jest.resetModules();
    jest.restoreAllMocks();
    //@ts-ignore
    utils.fetchUtils.mockResolvedValue(true);
    removeUserModal = new RemoveUserModal({
      login: 'login',
      guid: 'guid',
      getUsers: () => 'user',
    });
  });

  describe('render', () => {
    it('renders correctly', () => {
      const tree = renderer
        .create(
          <RemoveUserModal
            guid={'guid'}
            login={'login'}
            getUsers={'getUsers'}
          />,
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should call onClickUpdatePassword on click on up button', async () => {
      const component = renderer.create(
        <RemoveUserModal guid={'guid'} login={'login'} getUsers={'getUsers'} />,
      );
      const componentInstance: any = component.getInstance();
      jest.spyOn(componentInstance, 'onClickRemoveUser').mockResolvedValue('');
      component.root
        .findByProps({ id: 'remove-user-validate-button' })
        .props.onClick();
      expect(componentInstance.onClickRemoveUser).toHaveBeenCalledWith(
        'login',
        'guid',
      );
    });
  });

  describe('componentDidMount', () => {
    jest.mock('materialize-css');
    it('should call M.Modal.init with valid params', async () => {
      const modalInitMock = jest.fn().mockImplementation((modal, options) => {
        options.onCloseEnd();
      });
      M.Modal.init = modalInitMock;
      await removeUserModal.componentDidMount();
      expect(modalInitMock).toHaveBeenCalledWith(undefined, {
        onCloseEnd: expect.any(Function),
        inDuration: 250,
        outDuration: 250,
        opacity: 0.5,
        dismissible: true,
        startingTop: '4%',
        endingTop: '10%',
      });
    });
  });

  describe('onClickRemoveUser', () => {
    it('should call fetchUtils and call M.toast with green when fetch is in sucess', async () => {
      const toastMock = jest.fn();
      M.toast = toastMock;
      await removeUserModal.onClickRemoveUser('login', 'guid');
      expect(utils.fetchUtils).toHaveBeenCalledWith(
        '/api/users/remove?guid=guid',
        'DELETE',
        '{"login":"login"}',
      );
      expect(toastMock).toBeCalledWith({
        html: 'login a été supprimé avec succès',
        classes: 'green',
      });
    });
    it('should call fetchUtils and call M.toast with red when fetch is in error', async () => {
      const toastMock = jest.fn();
      M.toast = toastMock;
      //@ts-ignore
      utils.fetchUtils.mockRejectedValue();
      await removeUserModal.onClickRemoveUser('login', 'guid');
      expect(utils.fetchUtils).toHaveBeenCalledWith(
        '/api/users/remove?guid=guid',
        'DELETE',
        '{"login":"login"}',
      );
      expect(toastMock).toBeCalledWith({
        html: `login n'a pas été supprimé, une erreur s'est produite`,
        classes: 'red',
      });
    });
  });
});
