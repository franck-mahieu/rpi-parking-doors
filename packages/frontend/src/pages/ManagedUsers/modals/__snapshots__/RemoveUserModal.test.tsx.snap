// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoveUserModal render renders correctly 1`] = `
<span
  id="user-modal"
>
  <button
    className="waves-effect waves-light btn modal-trigger user-modal remove-button"
    data-target="remove-user-modal-login"
    title="Suppression de l'utilisateur"
  >
    Supprimer
  </button>
  <div
    className="modal"
    id="remove-user-modal-login"
  >
    <div
      className="modal-content"
    >
      <h4>
        Suppression de l'utilisateur
         
        login
      </h4>
      <p>
        Voulez vous vraiment supprimer l'utilisateur
         
        login
         ?
      </p>
    </div>
    <div
      className="modal-footer"
    >
      <button
        className="modal-close waves-effect waves-red btn-flat cancel-button"
      >
        Annuler
      </button>
      <button
        className="modal-close waves-effect waves-green btn-flat"
        id="remove-user-validate-button"
        onClick={[Function]}
      >
        Valider
      </button>
    </div>
  </div>
</span>
`;
